cpp/src/arrow/flight/server.cc

```diff
@@ -294,30 +295,97 @@ class FlightServiceImpl : public FlightService::Service {
   FlightServerBase* server_;
 };
 
-struct FlightServerBase::FlightServerBaseImpl {
-  std::unique_ptr<grpc::Server> server;
+struct FlightServerBase::Impl {
+  std::string address_;
+  std::unique_ptr<FlightServiceImpl> service_;
+  std::unique_ptr<grpc::Server> server_;
+
+  // Signal handling
+  std::vector<int> signals_;
+  std::vector<struct sigaction> old_signal_handlers_;
+  volatile int got_signal_;
```

Comment:
Why not atomic instead of volatile, that would make `GotSignal` concurrency safe.

Reply:
I don't think other methods in this class are thread-safe... That said, I can indeed use an atomic.